nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py

```diff
@@ -980,8 +979,9 @@ def __init__(
                 logging.info(f'Features saved to {self.features_pkl}')
 
         # wait until the master process writes to the processed data files
-        if torch.distributed.is_initialized():
-            torch.distributed.barrier()
+        if not master_device:
+            while features is None and not os.path.exists(self.features_pkl):
```

Comment:
There is a risk that a process will start reading `self.features_pkl` before master process finishes writing. To avoid such collision, I suggest to first write cache to a `tempfile.TemporaryFile` file and then rename temporary file to `self.features_pkl`.

Reply:
In my experience there is no guarantee that `master_device` (i.e. `global_rank==0`) will be the first to start. In case it starts last all processes will basically already have built their own cache, but the `master_device` will store it for no one actually needing it.  

Just a thought; Could a solution be to use distribute the preparation of cache over all processes? E.g. individual processes working on `global_rank` related chunks, (e.g. every `global_rank` line). And when all files are ready (i.e. `world_size` number of them) join them into a single cache, then reload. This obviously assumes a shared storage.

Reply:
Thank you for idea, @itzsimpl ! Your approach will certainly speed up features preparation in a multi-node setting, and it may improve performance in a single-node.

Although, for large datasets we usually use tarred datasets which are uploaded to a cluster before training. Validation and testing features are usually prepared fast and their preparation does not require a speed-up, in my opinion. And we already have multiprocessing https://github.com/NVIDIA/NeMo/blob/41f27a5e68934761b05dd373e7b288aa798858e5/nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py#L580, so a speed-up from your modification in a single-node setting will not be large.

@ekmb what do you think?

Reply:
I may be mistaken, but I vaguely remember reading in one of the chats here that multiprocessing in P&C is prone to hangups.

https://github.com/NVIDIA/NeMo/blob/41f27a5e68934761b05dd373e7b288aa798858e5/nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py#L129

https://github.com/NVIDIA/NeMo/blob/41f27a5e68934761b05dd373e7b288aa798858e5/nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py#L539-L541

Reply:
@itzsimpl a fair point. I suggest to make it a separate task and save it somewhere so that we would not forget about it.